### PR TITLE
Satellite console server gzip compression

### PR DESCRIPTION
**Pull request by crawter**
What: I have added new plugin to our bundle/minification process. Now it will generate not only css/js bundles, but also will produce compressed (gzip) version of it. Fonts also will be compressed. It will allow us to minify static content that user will upload. Also, satellite server is updated, to check, if browser is able to work with compressed file. If not - we will return original file. in other case - we will return generated by vue cli compressed version of a file.

Why: To minify static files and reduce website loading speed.

original files sizes :
 
File name:-----------------------Original size: -------Gzipped size:
dist/js/vendors.js ----------------556.08 KiB ---------179.79 KiB
dist/js/app.js----------------------360.42 KiB ----------94.22 KiB
dist/css/app.a34a807b.css----168.83 KiB-----------22.54 KiB
dist/fonts/font_bold.ttf----------261.6 KiB-------------115 KiB
dist/fonts/font_bold.ttf----------260.2 KiB-------------114.9 KiB
dist/fonts/font_bold.ttf----------263.6 KiB-------------114.9 KiB

So, in some cases we will reduce page size on ~ 1.2 MB

Please describe the tests:
 - Test 1:
 - Test 2:
 
Please describe the performance impact:
i do not gzip files each time when user requests it. All bundles with gzipped version will be generated only once. So performance impact should be really minor.

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

### CODE REVIEW

File: satellite/console/consoleweb/server.go [hunk 1]
```diff
@@ -408,3 +410,80 @@ func (s *Server) Run(ctx context.Context) (err error) {
...
+		if err != nil {
+			s.log.Info(err.Error())
+			fn.ServeHTTP(writer, request)
+			return
+		}
+
+		gzippedFilePath := filePath + ".gz"
+
+		// loading gzipped analog of original file
+		byt, err := ioutil.ReadFile(gzippedFilePath)
```
Comment (egonelbre) [inline satellite/console/consoleweb/server.go]: Rather than reading the file into memory, pass it into file handler with the appropriate modified path.

File: satellite/console/consoleweb/server.go [hunk 2]
```diff
@@ -408,3 +410,80 @@ func (s *Server) Run(ctx context.Context) (err error) {
...
+		}
+
+		filePath, err := s.getFilePathFromRequest(request)
+		if err != nil {
+			s.log.Info(err.Error())
+			fn.ServeHTTP(writer, request)
+			return
+		}
+
+		gzippedFilePath := filePath + ".gz"
```
[the remark this answers is not in the document]
Comment (crawter) [reply satellite/console/consoleweb/server.go]: this files are generated by vue-cli. Moreover, if gzipped version does not exist - we will return original file

File: satellite/console/consoleweb/server.go [hunk 3]
```diff
@@ -20,7 +22,7 @@ import (
 	"github.com/zeebo/errs"
 	"go.uber.org/zap"
 	"golang.org/x/sync/errgroup"
-	monkit "gopkg.in/spacemonkeygo/monkit.v2"
+	"gopkg.in/spacemonkeygo/monkit.v2"
```
[the remark this answers is not in the document]
Comment (crawter) [reply satellite/console/consoleweb/server.go]: doh, goimports tool always remove it :(

[the remark this answers is not in the document]
Comment (egonelbre) [reply satellite/console/consoleweb/server.go]: I'm not aware of goimports removing it. Are you sure you are not using some other formatter?

[the remark this answers is not in the document]
Comment (crawter) [reply satellite/console/consoleweb/server.go]: @egonelbre all our team members are using goland ide and have goimports installed. And all of us (me, @rikysya and @Barterio ) faced this issue with removing monkit allias

[the remark this answers is not in the document]
Comment (egonelbre) [reply satellite/console/consoleweb/server.go]: The monkit removal comes from Goland not goimports. You can disable it from Settings -> Editor Inspections -> redundant alias.

File: satellite/console/consoleweb/server.go [hunk 4]
```diff
@@ -129,74 +130,74 @@ func (s *Server) appHandler(w http.ResponseWriter, req *http.Request) {
...
 	header.Set("Content-Security-Policy", strings.Join(cspValues, "; "))
 
-	http.ServeFile(w, req, filepath.Join(s.config.StaticDir, "dist", "index.html"))
+	http.ServeFile(w, req, filepath.Join(server.config.StaticDir, "dist", "index.html"))
 }
 
 // bucketUsageReportHandler generate bucket usage report page for project
-func (s *Server) bucketUsageReportHandler(w http.ResponseWriter, req *http.Request) {
-	ctx := req.Context()
+func (server *Server) bucketUsageReportHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (mobyvb) [inline satellite/console/consoleweb/server.go]: In my opinion, `req` is better than `r`

File: satellite/console/consoleweb/server.go [hunk 5]
```diff
@@ -247,107 +248,107 @@ func (s *Server) createRegistrationTokenHandler(w http.ResponseWriter, req *http
...
 			zap.String("token", activationToken),
 			zap.Error(err))
 
-		s.serveError(w, req)
+		server.serveError(w, r)
 		return
 	}
 
-	http.ServeFile(w, req, filepath.Join(s.config.StaticDir, "static", "activation", "success.html"))
+	http.ServeFile(w, r, filepath.Join(server.config.StaticDir, "static", "activation", "success.html"))
```
Comment (egonelbre) [inline satellite/console/consoleweb/server.go]: Use the http.FileSystem created in constructor rather than `http.ServeFile`

File: satellite/console/consoleweb/server.go [hunk 6]
```diff
@@ -247,107 +248,107 @@ func (s *Server) createRegistrationTokenHandler(w http.ResponseWriter, req *http
...
-		err = s.service.ResetPassword(ctx, recoveryToken, password)
+		err = server.service.ResetPassword(ctx, recoveryToken, password)
 		if err != nil {
-			s.serveError(w, req)
+			server.serveError(w, r)
 			return
 		}
 
-		http.ServeFile(w, req, filepath.Join(s.config.StaticDir, "static", "resetPassword", "success.html"))
+		http.ServeFile(w, r, filepath.Join(server.config.StaticDir, "static", "resetPassword", "success.html"))
```
Comment (egonelbre) [inline satellite/console/consoleweb/server.go]: Is there a reason `success.html` isn't a template?

File: satellite/console/consoleweb/server.go [hunk 7]
```diff
@@ -247,107 +248,107 @@ func (s *Server) createRegistrationTokenHandler(w http.ResponseWriter, req *http
...
 	case http.MethodGet:
-		t, err := template.ParseFiles(filepath.Join(s.config.StaticDir, "static", "resetPassword", "resetPassword.html"))
+		t, err := template.ParseFiles(filepath.Join(server.config.StaticDir, "static", "resetPassword", "resetPassword.html"))
```
Comment (egonelbre) [inline satellite/console/consoleweb/server.go]: Let's try to avoid directly touching disk. But we can do a separate PR.

File: satellite/console/consoleweb/server.go [hunk 8]
```diff
@@ -394,17 +395,47 @@ func (s *Server) Run(ctx context.Context) (err error) {
...
+		// just return original file
+		if !isGzipSupported || !isNeededFormatToGzip {
+			fn.ServeHTTP(w, r)
+			return
+		}
+
+		w.Header().Set("Content-Type", mime.TypeByExtension(extension))
+		w.Header().Set("Content-Encoding", "gzip")
+
+		r.RequestURI += ".gz"
```
Comment (egonelbre) [inline satellite/console/consoleweb/server.go]: Requests shouldn't be modified. It can break things. https://golang.org/pkg/net/http/#Handler

File: satellite/console/consoleweb/server.go (hunk 4 above)
Comment (egonelbre) [reply satellite/console/consoleweb/server.go]: `r` is the convention.

File: satellite/console/consoleweb/server.go (hunk 5 above)
Comment (stefanbenten) [reply satellite/console/consoleweb/server.go]: Good idea yeah.

File: satellite/console/consoleweb/server.go (hunk 7 above)
Comment (stefanbenten) [inline satellite/console/consoleweb/server.go]: This would also lead to an parsing of the template, everytime someone is requesting that page. Cant we do that on startup instead?

File: satellite/console/consoleweb/server.go [hunk 9]
```diff
@@ -247,107 +248,107 @@ func (s *Server) createRegistrationTokenHandler(w http.ResponseWriter, req *http
...
-		http.Redirect(w, req, "https://storjlabs.atlassian.net/servicedesk/customer/portals", http.StatusSeeOther)
+		http.Redirect(w, r, "https://storjlabs.atlassian.net/servicedesk/customer/portals", http.StatusSeeOther)
 	}
 
 	// No need to check error as we anyway redirect user to support page
-	_ = s.service.RevokeResetPasswordToken(ctx, recoveryToken)
+	_ = server.service.RevokeResetPasswordToken(ctx, recoveryToken)
 
-	http.Redirect(w, req, "https://storjlabs.atlassian.net/servicedesk/customer/portals", http.StatusSeeOther)
+	http.Redirect(w, r, "https://storjlabs.atlassian.net/servicedesk/customer/portals", http.StatusSeeOther)
```
Comment (stefanbenten) [inline satellite/console/consoleweb/server.go]: Would it make sense to make this a more global variable for the URL?

Comment (stefanbenten) [reply satellite/console/consoleweb/server.go]: Something like `server.service.serviceURL`

Comment (egonelbre) [reply satellite/console/consoleweb/server.go]: This probably should be a config option with the default set to that.
